libdns.go

```diff
@@ -105,26 +147,87 @@ type ZoneLister interface {
-	Type  string
-	Name  string // partially-qualified (relative to zone)
+
+	// The `Type` field specifies the type of the record. Implementations may
+	// or may not support any given record type, and may support additional
+	// "private" record types with implementation-defined behavior.
+	//
+	// Examples: "A", "AAAA", "CNAME", "MX", "TXT"
+	Type string
+
+	// The `Name` field specifies the name of the record. It is partially
+	// qualified relative to the current zone. This field is called a "Label" by
+	// RFC7719. You may use "@" or the empty string "" to represent the root of
+	// the zone.
+	//
+	// Examples: "www", "", "@", "subdomain", "sub.subdomain"
+	//
+	// Invalid: "www.example.com.", "example.net." (fully-qualified)
+	//
+	// Valid, but probably doesn't do what you want: "www.example.com" (for a
+	// zone "example.net.", this refers to "www.example.com.example.net.")
+	Name string
+
+	// The `Value` field specifies the value of the record. This field should
+	// be formatted in the standard zone file syntax, but should omit any fields
+	// that are covered by other fields in this struct.
+	//
+	// Examples: (A)     "192.0.2.1"
+	//           (AAAA)  "2001:db8::1"
+	//           (CNAME) "example.com." (Even though the value is traditionally
+	//                   called the "target", it is included only in the `Value`
+	//                   field here.)
+	//           (MX)    "mail.example.com." (Note that this excludes the
+	//                   priority field!)
+	//	         (TXT)   "Hello, world!"
+	//	         (SRV)   "8080 example.com." (Note that this excludes the
+	//                   priority and weight fields, but includes the port.
+	//                   Also note that the target is included here, and not
+	//                   in the `Target` field.)
+	//	         (HTTPS) "alpn=h2,h3 port=443" (Note that this excludes the
+	//                   priority field and target fields.)
 	Value string
-	TTL   time.Duration
+
+	// The `TTL` field specifies the time-to-live of the record. This is
+	// represented in the DNS as an unsigned integral number of seconds, but
+	// is provided here as a time.Duration.
+	TTL time.Duration
 
 	// common, type-dependent record fields
-	Priority uint   // HTTPS, MX, SRV, and URI records
-	Weight   uint   // SRV and URI records
-	Target   string // HTTPS records
+
+	// The `Priority` field specifies the priority of the record. This field is
+	// only applicable for certain record types, but is mandatory for those
+	// types.
+	//
+	// Examples: (MX)    10 (Note that this is traditionally called the
+	//                   "preference" in the DNS)
+	//           (SRV)   10
+	//           (URI)   10
+	//           (HTTPS) 10
+	//           (SVCB)  10
+	Priority uint
+
+	// The `Weight` field specifies the weight of the record. This field is
+	// only applicable for certain record types, but is mandatory for those
+	// types.
+	//
+	// Examples: (SRV) 20
+	//           (URI) 20
+	Weight uint
```

Comment:
Considering that only `SRV` and `URI` records use the `Weight` field, the `SRV` records already force you to encode the port inside the `Value` field, and essentially nobody uses `URI` records, I wonder if it might make more sense to deprecate this field and recommend just encoding inside of the `Value` field. It's not great to need to encode the integer weight inside of the string `Value`, but you already have to do so for the port, so it shouldn't make much of a difference for the `libdns` interface users.

Reply:
I'm actually starting to agree with you on this. We can provide helper methods to parse out type-specific fields from the Value, but I am really starting to think that the Value should remain an opaque string...